scripts/generators/typescript.js

```diff
@@ -0,0 +1,186 @@
+"use strict";
+
+const t = require("../../packages/babel-types");
+const utils = require("./utils");
+
+let code = `// NOTE: This file is autogenerated. Do not modify.
+// See scripts/generators/typescript.js for script used.
+
+export interface Comment {
+  value: string;
+  start: number;
+  end: number;
+  loc: SourceLocation;
+  type: "BlockComment" | "LineComment";
+}
+
+export interface BlockComment extends Comment {
+  type: "BlockComment";
+}
+
+export interface LineComment extends Comment {
+  type: "LineComment";
+}
+
+export interface SourceLocation {
+  start: {
+    line: number;
+    column: number;
+  };
+
+  end: {
+    line: number;
+    column: number;
+  };
+}
+
+interface BaseNode {
+  leadingComments: ReadonlyArray<Comment> | null;
+  innerComments: ReadonlyArray<Comment> | null;
+  trailingComments: ReadonlyArray<Comment> | null;
+  start: number | null;
+  end: number | null;
+  loc: SourceLocation | null;
+  type: Node["type"];
+}
+
+export type Node = ${t.TYPES.sort().join(" | ")};\n\n`;
+
+//
+
+const lines = [];
+
+for (const type in t.NODE_FIELDS) {
+  const fields = t.NODE_FIELDS[type];
+  const fieldNames = sortFieldNames(Object.keys(t.NODE_FIELDS[type]), type);
+
+  const struct = ['type: "' + type + '";'];
+  const args = [];
+
+  fieldNames.forEach(fieldName => {
+    const field = fields[fieldName];
+    let typeAnnotation = utils.stringifyValidator(field.validate, "");
+
+    if (isNullable(field) && !hasDefault(field)) {
+      typeAnnotation += " | null";
+    }
+
+    if (areAllRemainingFieldsNullable(fieldName, fieldNames, fields)) {
+      args.push(
+        `${t.toBindingIdentifierName(fieldName)}${
+          isNullable(field) ? "?:" : ":"
+        } ${typeAnnotation}`
+      );
+    } else {
+      args.push(
+        `${t.toBindingIdentifierName(fieldName)}: ${typeAnnotation}${
+          isNullable(field) ? " | undefined" : ""
+        }`
+      );
+    }
+
+    if (t.isValidIdentifier(fieldName)) {
+      struct.push(`${fieldName}: ${typeAnnotation};`);
+    }
+  });
+
+  code += `export interface ${type} extends BaseNode {
+  ${struct.join("\n  ").trim()}
+}\n\n`;
+
+  // super and import are reserved words in JavaScript
+  if (type !== "Super" && type !== "Import") {
```

Comment:
You could generate something like:

```ts
function ${utils.toFunctionName(type)}Factory(${args.join(", ")}): ${type};
export { ${utils.toFunctionName(type)}Factory as ${utils.toFunctionName(type)} }
```

Reply:
@Kovensky I'm happy to do this after this PR is merged. This shouldn't block this PR.